README.md
# Docker-Terra

Dockerized Terra Node with Optional Bootstrap for Fast Syncing

* * *

[![CircleCI (all branches)](https://img.shields.io/circleci/build/github/RyanHendricks/docker-terra?token=e4ff41141d22bc71c6b2ad5118fd202d98dfb2dd&label=build&logo=circleci&logoColor=white)](https://circleci.com/gh/RyanHendricks/docker-terra)
[![Codacy grade](https://img.shields.io/codacy/grade/049014a16fd948edaaa0773c246517e9.svg?logo=codacy)](https://www.codacy.com?utm_source=github.com&utm_medium=referral&utm_content=RyanHendricks/docker-terra&utm_campaign=Badge_Grade)

[![MicroBadger Version](https://images.microbadger.com/badges/version/ryanhendricks/docker-terra.svg)](https://microbadger.com/images/ryanhendricks/docker-terra)
[![Docker Pulls](https://img.shields.io/docker/pulls/ryanhendricks/docker-terra.svg?logo=docker&logoColor=white)](https://hub.docker.com/r/ryanhendricks/docker-terra)
[![MicroBadger Layers (latest)](https://img.shields.io/microbadger/layers/ryanhendricks/docker-terra/latest.svg?logo=docker&logoColor=white)](https://microbadger.com/images/ryanhendricks/docker-terra)

* * *

## Prerequisites

- Docker
- Docker-Compose (optional)

## Quick Start

The image can be run without any configuration and defaults to mainnet

```bash

docker  run --rm -it -P --env SEEDS='447ba60df6fdc2466e1ab4c328e100d6bc5765f8@seed-only.terra-columbus-3.bas.network:26656,6be0856f6365559fdc2e9e97a07d609f754632b0@terra-columbus-3-seed.nodes.polychainlabs.com:26656,b416f0b04e2c71b8d76f993468352030e2dcf2a9@public-seed-node.columbus.certus.one:26656,87048bf71526fb92d73733ba3ddb79b7a83ca11e@public-seed.terra.dev:26656,b5205baf1d52b6f91afb0da7d7b33dcebc71755f@public-seed2.terra.dev:26656,535222fdb795df6653934f22b8e5f16fdfacc9f6@seed.terra.de-light.io:26656,bae08cc880c20aeda68a5a890a71a9b44ac73cb4@terra-seed-eu.chorus.one:28657,925ecc3de9e2ac65a203beb2333ced1a00c135ed@terra-seed-us.chorus.one:28657' ryanhendricks/docker-terra:latest
# Feel free to use an alternate seed node although without one the node will have issues starting

```

## Configuration

### Config.toml Parameters

- The config.toml is created dynamically when starting the container.
- All parameters specified in the standard config.toml file can be set using environmental variables with the same name as the config parameter but in all caps. See [entrypoint script](./scripts/entrypoint.sh) for more details.
- If left unset the default values will be used.
- Parameters can be set directly by modifying the config.toml portion of ./scripts/entrypoint.sh if you are cloning and building the image yourself.

### Environment Variables

```bash
# EXAMPLES

MONIKER=nonamenode
CHAIN_ID=columbus-3
GENESIS_URL=https://columbus-genesis.s3-ap-northeast-1.amazonaws.com/genesis.json
SEEDS=447ba60df6fdc2466e1ab4c328e100d6bc5765f8@seed-only.terra-columbus-3.bas.network:26656,6be0856f6365559fdc2e9e97a07d609f754632b0@terra-columbus-3-seed.nodes.polychainlabs.com:26656,b416f0b04e2c71b8d76f993468352030e2dcf2a9@public-seed-node.columbus.certus.one:26656,87048bf71526fb92d73733ba3ddb79b7a83ca11e@public-seed.terra.dev:26656,b5205baf1d52b6f91afb0da7d7b33dcebc71755f@public-seed2.terra.dev:26656,535222fdb795df6653934f22b8e5f16fdfacc9f6@seed.terra.de-light.io:26656,bae08cc880c20aeda68a5a890a71a9b44ac73cb4@terra-seed-eu.chorus.one:28657,925ecc3de9e2ac65a203beb2333ced1a00c135ed@terra-seed-us.chorus.one:28657
PERSISTENT_PEERS=e6325ba7c490ba371135c9f3fcead66da1bd8cf1@terra-sentry01.dokia.cloud:26656,dba5defd7b120937da37aea7f37d06870637558d@terra-sentry02.dokia.cloud:26656,eb4ce12133c450ba6665e06309570ea2843e21d8@167.86.104.33:26656,7277be5ce17d60cf26c92a7cafbb9fc7da7f2be5@51.38.103.128:26656,1cb3e13efe7ca25fb68249169a15e85e53c3b3e9@terra-main.peer.nodeateam.kr:26656,46bc5183ef3b6ea9ffa84df16d6a5aa4a642427a@node.terra.forbole.com:26656,


```

You can set ENV variables either in a docker-compose file or in the docker run command if running the container directly. If left unchanged they will default to the standard value except for the following:

- MONIKER
  - defaults to "nonamenode"
- CHAIN_ID
  - defaults to columbus-3
- GENESIS_URL
  - defaults to columbus-3 github [genesis file url](https://columbus-genesis.s3-ap-northeast-1.amazonaws.com/genesis.json)

### Bootstrapping

  The entrypoint script allows for the node to be bootstrapped upon creation. This drastically reduces the time required to fully sync with the network. The start up time is increased relative to the amount of time it takes the host machine to download ~20GB of data and extract it into the data directory.

  Set the ENV variable `BOOTSTRAP=TRUE` to enable bootstrapping.

## Build

The following command will build the image.

```bash
docker build --rm -f Dockerfile -t docker-terra:latest .
```

## Running

### Mainnet

```sh
docker-compose up -d --build
```

### Testnet

```sh
docker-compose docker-compose-testnet.yml up -d --build
```

## Supervisor

The image uses Supervisor to run both terrad and gaiacli simultaneously at container runtime. Supervisor also restarts either process should it fail for some reason.

### Gaiad

After starting the container you can check the status here: <http://127.0.0.1:26657/status>.

or from the terminal

```bash
curl -X GET \
  http://127.0.0.1:26657/status? \
  -H 'cache-control: no-cache'
```

## Gaiacli Rest-Server

Supervisor starts the rest-server with the following command:

```bash
gaiacli rest-server --trust-node --cors * --home $GAIAD_HOME --laddr tcp://0.0.0.0:1317

```

You can verify that the rest-server is running using the following example

```bash
curl -X GET \
  http://127.0.0.1:1317/blocks/latest \
  -H 'cache-control: no-cache'
```

## NOTES

- The current SEEDS (nodes I am running and supplying here since the ones from terra/launch repo all are not working as of this update) may not be around forever so consider overriding the defaults. If the badges above have do not have block numbers for either chain that means the seed nodes are no longer with us.
- You probably should not run a validator with this setup.

## Contributing

[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?)](http://makeapullrequest.com)

### Contributors

[![Keybase PGP](https://img.shields.io/keybase/pgp/ryanhendricks.svg?label=keybase&logo=keybase&logoColor=white)](https://keybase.io/ryanhendricks)

## License

![GitHub](https://img.shields.io/github/license/ryanhendricks/docker-terra.svg)

docker run -it --rm -v "$(pwd)":/usr/local/src/your-app ferrarimarco/github-changelog-generator -u TokenUnion -p union-marketplace --token e19f446c77a0f02675815ff42a21e72248546824


CHANGELOG.md
# Changelog

## [v1.0.0](https://github.com/RyanHendricks/docker-terra/tree/v1.0.0) (2020-01-31)

[Full Changelog](https://github.com/RyanHendricks/docker-terra/compare/columbus-3...v1.0.0)

## [columbus-3](https://github.com/RyanHendricks/docker-terra/tree/columbus-3) (2019-12-26)

[Full Changelog](https://github.com/RyanHendricks/docker-terra/compare/4b6fe50a03427f59961d1c224608c741271f2fd2...columbus-3)


scripts/entrypoint.sh
#!/bin/sh


# exit script on any error
set -e
echo "setting up initial configurations"

if [ ! -f "$TERRAD_HOME/config/config.toml" ];
then

  terrad init "${MONIKER:-nonamenode}" --home="${TERRAD_HOME:-/.terrad}"
  
  cd "$TERRAD_HOME/config"

  rm genesis.json
  rm config.toml

  if [ ! -z "$GENESIS_URL" ]; then
    wget "$GENESIS_URL"
  else
    # Mainnet
    wget https://columbus-genesis.s3-ap-northeast-1.amazonaws.com/genesis.json
  fi

# if [ -f "$TERRAD_HOME/config/genesis.json" ]; then
#   rm genesis.json
# fi
# if [ -f "$TERRAD_HOME/config/config.toml" ]; then
#   rm config.toml
# fi


cat > config.toml << EOF
# This is a TOML config file.
# For more information, see https://github.com/toml-lang/toml

##### main base config options #####

# TCP or UNIX socket address of the ABCI application,
# or the name of an ABCI application compiled in with the Tendermint binary
proxy_app = "${PROXY_APP:-tcp://0.0.0.0}:${PROXY_APP_PORT:-26658}"

# A custom human readable name for this node
moniker = "${MONIKER:-moniker}"

# If this node is many blocks behind the tip of the chain, FastSync
# allows them to catchup quickly by downloading blocks in parallel
# and verifying their commits
fast_sync = ${FAST_SYNC:-true}

# Database backend: leveldb | memdb | cleveldb
db_backend = "${DB_BACKEND:-goleveldb}"

# Database directory
db_dir = "${DB_DIR:-data}"

# Output level for logging, including package level options
log_level = "${LOG_LEVEL:-main:info,state:info,*:error}"

log_format = "${LOG_FORMAT:-plain}"

##### additional base config options #####

# Path to the JSON file containing the initial validator set and other meta data
genesis_file = "${GENESIS_FILE:-config/genesis.json}"

# Path to the JSON file containing the private key to use as a validator in the consensus protocol
priv_validator_key_file = "${PRIV_VALIDATOR_KEY_FILE:-config/priv_validator_key.json}"

# Path to the JSON file containing the last sign state of a validator
priv_validator_state_file = "${PRIV_VALIDATOR_KEY_FILE:-data/priv_validator_state.json}"

# TCP or UNIX socket address for Tendermint to listen on for
# connections from an external PrivValidator process
priv_validator_laddr = "${PRIV_VALIDATOR_LADDR:-}"

# Path to the JSON file containing the private key to use for node authentication in the p2p protocol
node_key_file = "${NODE_KEY_FILE:-config/node_key.json}"

# Mechanism to connect to the ABCI application: socket | grpc
abci = "${ABCI:-socket}"

# TCP or UNIX socket address for the profiling server to listen on
prof_laddr = "${PROF_LADDR:-localhost:6060}"

# If true, query the ABCI app on connecting to a new peer
# so the app can decide if we should keep the connection or not
filter_peers = ${FILTER_PEERS:-false}

##### advanced configuration options #####

##### rpc server configuration options #####
[rpc]

# TCP or UNIX socket address for the RPC server to listen on
laddr = "${RPC_LADDR:-tcp://0.0.0.0}:${RPC_PORT:-26657}"


# A list of origins a cross-domain request can be executed from
# Default value '[]' disables cors support
# Use '["*"]' to allow any origin
cors_allowed_origins = ["*"]

# A list of methods the client is allowed to use with cross-domain requests
cors_allowed_methods = ["HEAD", "GET", "POST"]

# A list of non simple headers the client is allowed to use with cross-domain requests
cors_allowed_headers = ["Origin", "Accept", "Content-Type", "X-Requested-With", "X-Server-Time"]



# TCP or UNIX socket address for the gRPC server to listen on
# NOTE: This server only supports /broadcast_tx_commit
grpc_laddr = "${GRPC_LADDR:-}"

# Maximum number of simultaneous connections.
# Does not include RPC (HTTP&WebSocket) connections. See max_open_connections
# If you want to accept more significant number than the default, make sure
# you increase your OS limits.
# 0 - unlimited.
# Should be < {ulimit -Sn} - {MaxNumInboundPeers} - {MaxNumOutboundPeers} - {N of wal, db and other open files}
# 1024 - 40 - 10 - 50 = 924 = ~900
grpc_max_open_connections = ${GRPC_MAX_OPEN_CONNECTIONS:-900}

# Activate unsafe RPC commands like /dial_seeds and /unsafe_flush_mempool
unsafe = ${UNSAFE:-false}

# Maximum number of simultaneous connections (including WebSocket).
# Does not include gRPC connections. See grpc_max_open_connections
# If you want to accept more significant number than the default, make sure
# you increase your OS limits.
# 0 - unlimited.
# Should be < {ulimit -Sn} - {MaxNumInboundPeers} - {MaxNumOutboundPeers} - {N of wal, db and other open files}
# 1024 - 40 - 10 - 50 = 924 = ~900
max_open_connections = ${MAX_OPEN_CONNECTIONS:-900}

# Maximum number of unique clientIDs that can /subscribe
# If you're using /broadcast_tx_commit, set to the estimated maximum number
# of broadcast_tx_commit calls per block.
max_subscription_clients = ${MAX_SUBSCRIPTION_CLIENTS:-100}

# Maximum number of unique queries a given client can /subscribe to
# If you're using GRPC (or Local RPC client) and /broadcast_tx_commit, set to
# the estimated # maximum number of broadcast_tx_commit calls per block.
max_subscriptions_per_client = ${MAX_SUBSCRIPTION_PER_CLIENT:-5}

# How long to wait for a tx to be committed during /broadcast_tx_commit.
# WARNING: Using a value larger than 10s will result in increasing the
# global HTTP write timeout, which applies to all connections and endpoints.
# See https://github.com/tendermint/tendermint/issues/3435
timeout_broadcast_tx_commit = "${TIMEOUT_BROADCAST_TX_COMMIT:-10s}"


# Maximum size of request body, in bytes
max_body_bytes = ${MAX_SIZE_REQUEST_BODY:-1000000}

# Maximum size of request header, in bytes
max_header_bytes = ${MAX_SIZE_REQUEST_HEADER:-1048576}

# The name of a file containing certificate that is used to create the HTTPS server.
# If the certificate is signed by a certificate authority,
# the certFile should be the concatenation of the server's certificate, any intermediates,
# and the CA's certificate.
# NOTE: both tls_cert_file and tls_key_file must be present for Tendermint to create HTTPS server. Otherwise, HTTP server is run.
tls_cert_file = "${TLS_CERT_FILE:-}"

# The name of a file containing matching private key that is used to create the HTTPS server.
# NOTE: both tls_cert_file and tls_key_file must be present for Tendermint to create HTTPS server. Otherwise, HTTP server is run.
tls_key_file = "${TLS_KEY_FILE:-}"


##### peer to peer configuration options #####
[p2p]

# Address to listen for incoming connections
laddr = "${P2P_LADDR:-tcp://0.0.0.0}:${P2P_PORT:-26656}"

# Address to advertise to peers for them to dial
# If empty, will use the same port as the laddr,
# and will introspect on the listener or use UPnP
# to figure out the address.
external_address = "${EXTERNAL_ADDRESS:-}"

# Comma separated list of seed nodes to connect to
seeds = "${SEEDS:-}"

# Comma separated list of nodes to keep persistent connections to
persistent_peers = "${PERSISTENT_PEERS:-}"

# UPNP port forwarding
upnp = ${UPNP:-false}

# Path to address book
addr_book_file = "${ADDR_BOOK_FILE:-config/addrbook.json}"

# Set true for strict address routability rules
# Set false for private or local networks
addr_book_strict = ${ADDR_BOOK_STRICT:-true}

# Maximum number of inbound peers
max_num_inbound_peers = ${MAX_NUM_INBOUND_PEERS:-40}

# Maximum number of outbound peers to connect to, excluding persistent peers
max_num_outbound_peers = ${MAX_NUM_OUTBOUND_PEERS:-10}

# Time to wait before flushing messages out on the connection
flush_throttle_timeout = "${FLUSH_THROTTLE_TIMEOUT:-100ms}"

# Maximum size of a message packet payload, in bytes
max_packet_msg_payload_size = ${MAX_PACKET_MSG_PAYLOAD_SIZE:-1024}

# Rate at which packets can be sent, in bytes/second
send_rate = ${SEND_RATE:-5120000}

# Rate at which packets can be received, in bytes/second
recv_rate = ${RECV_RATE:-5120000}

# Set true to enable the peer-exchange reactor
pex = ${PEX:-true}

# Seed mode, in which node constantly crawls the network and looks for
# peers. If another node asks it for addresses, it responds and disconnects.
#
# Does not work if the peer-exchange reactor is disabled.
seed_mode = ${SEED_MODE:-false}

# Comma separated list of peer IDs to keep private (will not be gossiped to other peers)
private_peer_ids = "${PRIVATE_PEER_IDS:-}"

# Toggle to disable guard against peers connecting from the same ip.
allow_duplicate_ip = ${ALLOW_DUPLICATE_IP:-false}

# Peer connection configuration.
handshake_timeout = "${HANDSHAKE_TIMEOUT:-20s}"
dial_timeout = "${DIAL_TIMEOUT:-3s}"

##### mempool configuration options #####
[mempool]

recheck = ${RECHECK:-true}
broadcast = ${BROADCAST:-true}
wal_dir = "${WAL_DIR}"

# size of the mempool
size = ${SIZE_OF_MEMPOOL:-5000}

# This only accounts for raw transactions (e.g. given 1MB transactions and
# max_txs_bytes=5MB, mempool will only accept 5 transactions).
max_txs_bytes = ${MAX_TXS_BYTES:-1073741824}


# size of the cache (used to filter transactions we saw earlier)
cache_size = ${CACHE_SIZE:-10000}


# Maximum size of a single transaction.
max_tx_bytes = ${MAX_TX_BYTES:-1048576}

##### fast sync configuration options #####
[fastsync]

# Fast Sync version to use:
#   1) "v0" (default) - the legacy fast sync implementation
#   2) "v1" - refactor of v0 version for better testability
version = "${FAST_SYNC_VERSION:-v0}"



##### consensus configuration options #####
[consensus]

wal_file = "${WAL_FILE:-data/cs.wal/wal}"

timeout_propose = "${TIMEOUT_PROPOSE:-3s}"
timeout_propose_delta = "${TIMEOUT_PROPOSE_DELTA:-500ms}"
timeout_prevote = "${TIMEOUT_PREVOTE:-1s}"
timeout_prevote_delta = "${TIMEOUT_PREVOTE_DELTA:-500ms}"
timeout_precommit = "${TIMEOUT_PRECOMMIT:-1s}"
timeout_precommit_delta = "${TIMEOUT_PRECOMMIT_DELTA:-500ms}"
timeout_commit = "${TIMEOUT_COMMIT:-5s}"


# Make progress as soon as we have all the precommits (as if TimeoutCommit = 0)
skip_timeout_commit = ${SKIP_TIMEOUT_COMMIT:-false}

# EmptyBlocks mode and possible interval between empty blocks
create_empty_blocks = ${CREATE_EMPTY_BLOCKS:-true}
create_empty_blocks_interval = "${CREATE_EMPTY_BLOCKS_INTERVAL:-0s}"

# Reactor sleep duration parameters
peer_gossip_sleep_duration = "${PEER_GOSSIP_SLEEP_DURATION:-100ms}"
peer_query_maj23_sleep_duration = "${PEER_QUERY_MAJ23_SLEEP_DURATION:-2s}"

##### transactions indexer configuration options #####
[tx_index]

# What indexer to use for transactions
#
# Options:
#   1) "null" (default)
#   2) "kv" - the simplest possible indexer, backed by key-value storage (defaults to levelDB; see DBBackend).
indexer = "${INDEXER_SELECTION:-kv}"

# Comma-separated list of tags to index (by default the only tag is "tx.hash")
#
# You can also index transactions by height by adding "tx.height" tag here.
#
# It's recommended to index only a subset of tags due to possible memory
# bloat. This is, of course, depends on the indexer's DB and the volume of
# transactions.
# index_tags = "${INDEX_TAGS:-}"

# When set to true, tells indexer to index all tags (predefined tags:
# "tx.hash", "tx.height" and all tags from DeliverTx responses).
#
# Note this may be not desirable (see the comment above). IndexTags has a
# precedence over IndexAllTags (i.e. when given both, IndexTags will be
# indexed).
index_all_tags = ${INDEX_ALL_TAGS:-true}

##### instrumentation configuration options #####
[instrumentation]

# When true, Prometheus metrics are served under /metrics on
# PrometheusListenAddr.
# Check out the documentation for the list of available metrics.
prometheus = ${PROMETHEUS:-false}

# Address to listen for Prometheus collector(s) connections
prometheus_listen_addr = ":${PROMETHEUS_LISTEN_ADDR:-26660}"

# Maximum number of simultaneous connections.
# If you want to accept more significant number than the default, make sure
# you increase your OS limits.
# 0 - unlimited.
max_open_connections = ${MAX_OPEN_CONNECTIONS:-3}

# Instrumentation namespace
namespace = "${INSTRUMENTATION_NAMESPACE:-tendermint}"


EOF



cd "$TERRAD_HOME"

  if [ "$BOOTSTRAP" == "TRUE" ]; then
    echo "bootstrapping... this will take some time."
    wget 	http://quicksync.chainlayer.io/terra/colombus-3.20200128.0205.tar.lz4	
    lz4 -d -v --rm colombus-3.20200128.0205.tar.lz4 | tar xf -
  else
      echo "bootstrap ENV variable != TRUE -->  syncing chain from genesis..."
  fi

fi


if [ ! -z "$LCD_PORT" ]; then
    rm /etc/supervisor/conf.d/supervisor-terracli.conf
    cd /etc/supervisor/conf.d/

cat > supervisor-terracli.conf << EOF
[program:terracli]
command=terracli rest-server --laddr tcp://0.0.0.0:${LCD_PORT:-1317} --home=${TERRAD_HOME:-/.terrad} --trust-node --node=${RPC_LADDR:-tcp://0.0.0.0}:${RPC_PORT:-26657}
redirect_stderr=true
EOF

fi

exec supervisord --nodaemon --configuration /etc/supervisor/supervisord.conf


# terrad start --home=$TERRAD_HOME
